services/federator/test/unit/Test/Federator/Validation.hs

```diff
@@ -57,46 +78,135 @@ tests =
+      x -> assertFailure $ "expected IFederationDeniedByRemote error, got " <> show x
 
 validateDomainAllowListSuccess :: TestTree
 validateDomainAllowListSuccess =
-  testCase "should give parsed domain if in the allow list" $
-    -- removing evalMock @Remote doesn't seem to work, but why?
-    runM . evalMock @Remote @IO $ do
-      let domain = Domain "hello.world"
-      let settings = settingsWithAllowList [domain]
-      res :: Either InwardError Domain <- Polysemy.runError . Polysemy.runReader settings $ validateDomain ("hello.world" :: Text)
-      embed $ assertEqual "validateDomain should give 'hello.world' as domain" (Right domain) res
+  testCase "should give parsed domain if in the allow list" $ do
+    exampleCert <- BS.readFile "test/resources/unit/localhost.example.com.pem"
+    let domain = Domain "localhost.example.com"
+        settings = settingsWithAllowList [domain]
+        res :: Either InwardError Domain =
+          run
+            . Polysemy.runError
+            . mockDiscoveryTrivial
+            . Polysemy.runReader settings
+            $ validateDomain (Just exampleCert) (domainText domain)
+    assertEqual "validateDomain should give 'localhost.example.com' as domain" (Right domain) res
+
+validateDomainCertMissing :: TestTree
+validateDomainCertMissing =
+  testCase "should fail if no client certificate is provided" $ do
+    let res :: Either InwardError Domain =
+          run . Polysemy.runError
+            . mockDiscoveryTrivial
+            . Polysemy.runReader noClientCertSettings
+            $ validateDomain Nothing "foo.example.com"
+    case res of
+      Left (InwardError IAuthenticationFailed _) -> pure ()
+      x -> assertFailure $ "expected IAuthenticationFailed error, got " <> show x
+
+validateDomainCertInvalid :: TestTree
+validateDomainCertInvalid =
+  testCase "should fail if the client certificate is invalid" $ do
+    let res :: Either InwardError Domain =
+          run . Polysemy.runError
+            . mockDiscoveryTrivial
+            . Polysemy.runReader noClientCertSettings
+            $ validateDomain (Just "not a certificate") "foo.example.com"
+    case res of
+      Left (InwardError IAuthenticationFailed _) -> pure ()
+      x -> assertFailure $ "expected IAuthenticationFailed error, got " <> show x
+
+validateDomainCertWrongDomain :: TestTree
+validateDomainCertWrongDomain =
+  testCase "should fail if the client certificate has a wrong domain" $ do
+    exampleCert <- BS.readFile "test/resources/unit/localhost.example.com.pem"
+    let res :: Either InwardError Domain =
+          run . Polysemy.runError
+            . mockDiscoveryTrivial
+            . Polysemy.runReader noClientCertSettings
+            $ validateDomain (Just exampleCert) "foo.example.com"
+    case res of
+      Left (InwardError IAuthenticationFailed _) -> pure ()
+      x -> assertFailure $ "expected IAuthenticationFailed error, got " <> show x
```

Comment:
maybe create a little helper function similar to `expectErr` in this file to not duplicate these three lines of case statements for expected errors?

Reply:
I just replaced this with asserting more precisely on the errors.